File: analytical_engine/java/graphx-on-graphscope/src/main/java/com/alibaba/graphscope/utils/GraphXProxy.java
package com.alibaba.graphscope.utils;

import com.alibaba.graphscope.conf.GraphXConf;
import com.alibaba.graphscope.ds.Vertex;
import com.alibaba.graphscope.factory.GraphXFactory;
import com.alibaba.graphscope.fragment.IFragment;
import com.alibaba.graphscope.graph.EdgeContextImpl;
import com.alibaba.graphscope.graph.GraphXVertexIdManager;
import com.alibaba.graphscope.graph.GraphxEdgeManager;
import com.alibaba.graphscope.graph.VertexDataManager;
import com.alibaba.graphscope.graphx.GSEdgeTriplet;
import com.alibaba.graphscope.mm.MessageStore;
import com.alibaba.graphscope.parallel.DefaultMessageManager;
import com.alibaba.graphscope.parallel.message.DoubleMsg;
import com.alibaba.graphscope.parallel.message.LongMsg;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;
import org.apache.spark.graphx.EdgeTriplet;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import scala.Function1;
import scala.Function2;
import scala.Function3;
import scala.Tuple2;
import scala.collection.Iterator;

public class GraphXProxy<VD, ED, MSG_T> {

    private static Logger logger = LoggerFactory.getLogger(GraphXProxy.class.getName());
    private static String SPARK_LAUNCHER_OUTPUT = "spark_laucher_output";
    private static final int chunkSize = 1024;
    /**
     * User vertex program: vprog: (VertexId, VD, A) => VD
     */
    private Function3<Long, VD, MSG_T, VD> vprog;
    /**
     * EdgeTriplet[VD, ED] => Iterator[(VertexId, A)]
     */
    private Function1<EdgeTriplet<VD, ED>, Iterator<Tuple2<Long, MSG_T>>> sendMsg;
    /**
     * (A, A) => A)
     */
    private Function2<MSG_T, MSG_T, MSG_T> mergeMsg;
    private GraphXVertexIdManager idManager;
    private VertexDataManager<VD> vertexDataManager;
    private MessageStore<MSG_T,VD> inComingMessageStore, outgoingMessageStore;
//    private EdgeContextImpl<VD, ED, MSG_T> edgeContext;
    private GSEdgeTriplet<VD,ED> edgeTriplet;
    private GSEdgeTriplet<VD,ED>[] edgeTriplets;
    private GraphXConf<VD, ED, MSG_T> conf;
    private GraphxEdgeManager<VD, ED, MSG_T> edgeManager;
    private DefaultMessageManager messageManager;
    private IFragment<Long, Long, VD, ED> graphxFragment; // different from c++ frag
    private MSG_T initialMessage;
    private ExecutorService executorService;
    private int numCores;

    public GraphXProxy(GraphXConf<VD, ED, MSG_T> conf, Function3<Long, VD, MSG_T, VD> vprog,
        Function1<EdgeTriplet<VD, ED>, Iterator<Tuple2<Long, MSG_T>>> sendMsg,
        Function2<MSG_T, MSG_T, MSG_T> mergeMsg, int numCores) {
        this.numCores = numCores;
        this.conf = conf;
        this.vprog = vprog;
        this.sendMsg = sendMsg;
        this.mergeMsg = mergeMsg;
        //create all objects here, but not initialized, initialize after main invoke and graph got.
        this.idManager = GraphXFactory.createIdManager(conf);
        this.vertexDataManager = GraphXFactory.createVertexDataManager(conf);
        //fixme: parallel
        this.inComingMessageStore = GraphXFactory.createParallelMessageStore(conf);
        this.outgoingMessageStore = GraphXFactory.createParallelMessageStore(conf);
//        this.edgeContext = GraphXFactory.createEdgeContext(conf);
        this.edgeTriplet = GraphXFactory.createEdgeTriplet(conf);
        this.edgeTriplets = new GSEdgeTriplet[numCores];
        for (int i = 0; i < numCores; ++i){
            this.edgeTriplets[i] = GraphXFactory.createEdgeTriplet(conf);
        }
        this.edgeManager = GraphXFactory.createEdgeManager(conf, idManager, vertexDataManager);
        executorService = Executors.newFixedThreadPool(numCores);
    }

    public void init(IFragment<Long, Long, VD, ED> fragment, DefaultMessageManager messageManager,
        MSG_T initialMessage) {
        this.graphxFragment = fragment;
        this.messageManager = messageManager;
        this.initialMessage = initialMessage;

        idManager.init(graphxFragment);
        vertexDataManager.init(graphxFragment);
        inComingMessageStore.init(graphxFragment, idManager,vertexDataManager, mergeMsg);
        outgoingMessageStore.init(graphxFragment,  idManager, vertexDataManager,mergeMsg);
//        edgeContext.init(outgoingMessageStore);
        //edgeTriplet no initialization
        edgeManager.init(graphxFragment, numCores);
    }

    public void PEval() {
        Vertex<Long> vertex = FFITypeFactoryhelper.newVertexLong();
        long innerVerticesNum = this.graphxFragment.getInnerVerticesNum();
        for (long lid = 0; lid < innerVerticesNum; ++lid) {
            vertexDataManager.setVertexData(lid,
                vprog.apply(idManager.lid2Oid(lid), vertexDataManager.getVertexData(lid),
                    initialMessage));
        }
        logger.info("[PEval] Finish vprog for frag {}", graphxFragment.fid());

        for (long lid = 0; lid < innerVerticesNum; ++lid) {
//            edgeContext.setSrcValues(idManager.lid2Oid(lid), lid,
//                vertexDataManager.getVertexData(lid));
            edgeTriplet.setSrcOid(idManager.lid2Oid(lid), vertexDataManager.getVertexData(lid));
            edgeManager.iterateOnEdges(lid, edgeTriplet, sendMsg, outgoingMessageStore);
        }
        logger.info("[PEval] Finish iterate edges for frag {}", graphxFragment.fid());
        outgoingMessageStore.flushMessage(messageManager);
        //messages to self are cached locally.
    }

    public void ParallelPEval() {
        {
            int innerVerticesNum = (int) this.graphxFragment.getInnerVerticesNum();
            AtomicInteger atomicInteger = new AtomicInteger(0);
            CountDownLatch countDownLatch = new CountDownLatch(numCores);
            int originEnd = innerVerticesNum;
            for (int tid = 0; tid < numCores; ++tid) {
                executorService.execute(
                    () -> {
                        while (true) {
                            int curBegin =
                                Math.min(atomicInteger.getAndAdd(chunkSize), originEnd);
                            int curEnd = Math.min(curBegin + chunkSize, originEnd);
                            if (curBegin >= originEnd) {
                                break;
                            }
                            try {
                                for (long lid = curBegin; lid < curEnd; ++lid) {
                                    vertexDataManager.setVertexData(lid,
                                        vprog.apply(idManager.lid2Oid(lid), vertexDataManager.getVertexData(lid),
                                            initialMessage));
                                }
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        }
                        countDownLatch.countDown();
                    });
            }
            try {
                countDownLatch.await();
            } catch (Exception e) {
                e.printStackTrace();
                executorService.shutdown();
            }
        }
        {
            int innerVerticesNum = (int) this.graphxFragment.getInnerVerticesNum();
            AtomicInteger atomicInteger = new AtomicInteger(0);
            CountDownLatch countDownLatch = new CountDownLatch(numCores);
            int originEnd = innerVerticesNum;
            for (int tid = 0; tid < numCores; ++tid) {
                GSEdgeTriplet<VD,ED> threadTriplet = edgeTriplets[tid];
                int finalTid = tid;
                executorService.execute(
                    () -> {
                        while (true) {
                            int curBegin =
                                Math.min(atomicInteger.getAndAdd(chunkSize), originEnd);
                            int curEnd = Math.min(curBegin + chunkSize, originEnd);
                            if (curBegin >= originEnd) {
                                break;
                            }
                            try {
                                for (long lid = curBegin; lid < curEnd; ++lid) {
                                    threadTriplet.setSrcOid(idManager.lid2Oid(lid), vertexDataManager.getVertexData(lid));
                                    edgeManager.iterateOnEdgesParallel(finalTid, lid, threadTriplet, sendMsg, outgoingMessageStore);
                                }
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        }
                        countDownLatch.countDown();
                    });
            }
            try {
                countDownLatch.await();
            } catch (Exception e) {
                e.printStackTrace();
                executorService.shutdown();
            }
        }
        outgoingMessageStore.flushMessage(messageManager);
        //messages to self are cached locally.
    }

    public void IncEval() {
        outgoingMessageStore.swap(inComingMessageStore);
        Vertex<Long> receiveVertex = FFITypeFactoryhelper.newVertexLong();
        boolean outerMsgReceived = receiveMessage(receiveVertex);
        long innerVerticesNum = this.graphxFragment.getInnerVerticesNum();

        inComingMessageStore.clear();
        outgoingMessageStore.swap(inComingMessageStore);
        outgoingMessageStore.clear();
        if (outerMsgReceived || inComingMessageStore.hasMessages()) {
            for (long lid = 0; lid < innerVerticesNum; ++lid) {
                if (inComingMessageStore.messageAvailable(lid)) {
                    vertexDataManager.setVertexData(lid, vprog.apply(idManager.lid2Oid(lid),
                        vertexDataManager.getVertexData(lid),
                        inComingMessageStore.getMessage(lid)));
                }
            }

            int sendMsgCnt = 0;
            //after running vprog, we now send msg and merge msg
            for (long lid = 0; lid < innerVerticesNum; ++lid) {
                if (inComingMessageStore.messageAvailable(lid)) {
                    edgeTriplet.setSrcOid(idManager.lid2Oid(lid), vertexDataManager.getVertexData(lid));
                    edgeManager.iterateOnEdges(lid, edgeTriplet, sendMsg, outgoingMessageStore);
                    sendMsgCnt += 1;
                }
            }
            logger.info("frag {} vprog runned for {} times", graphxFragment.fid(), sendMsgCnt);

            //FIXME: flush message
            outgoingMessageStore.flushMessage(messageManager);
        } else {
            logger.info("Frag {} No message received", graphxFragment.fid());
        }
    }

    public void ParallelIncEval() {
        Vertex<Long> receiveVertex = FFITypeFactoryhelper.newVertexLong();
        boolean outerMsgReceived = receiveMessage(receiveVertex);
        long innerVerticesNum = this.graphxFragment.getInnerVerticesNum();

        inComingMessageStore.clear();
        outgoingMessageStore.swap(inComingMessageStore);
        outgoingMessageStore.clear();
        if (outerMsgReceived || inComingMessageStore.hasMessages()) {
            {
                AtomicInteger atomicInteger = new AtomicInteger(0);
                CountDownLatch countDownLatch = new CountDownLatch(numCores);
                int originEnd = (int) innerVerticesNum;
                for (int tid = 0; tid < numCores; ++tid) {
                    executorService.execute(
                        () -> {
                            while (true) {
                                int curBegin =
                                    Math.min(atomicInteger.getAndAdd(chunkSize), originEnd);
                                int curEnd = Math.min(curBegin + chunkSize, originEnd);
                                if (curBegin >= originEnd) {
                                    break;
                                }
                                try {
                                    for (long lid = curBegin; lid < curEnd; ++lid) {
                                        if (inComingMessageStore.messageAvailable(lid)) {
                                            vertexDataManager.setVertexData(lid, vprog.apply(idManager.lid2Oid(lid),
                                                vertexDataManager.getVertexData(lid),
                                                inComingMessageStore.getMessage(lid)));
                                        }
                                    }
                                } catch (Exception e) {
                                    e.printStackTrace();
                                }
                            }
                            countDownLatch.countDown();
                        });
                }
                try {
                    countDownLatch.await();
                } catch (Exception e) {
                    e.printStackTrace();
                    executorService.shutdown();
                }
            }

            {
                AtomicInteger atomicInteger = new AtomicInteger(0);
                CountDownLatch countDownLatch = new CountDownLatch(numCores);
                int originEnd = (int) innerVerticesNum;
                for (int tid = 0; tid < numCores; ++tid) {
                    GSEdgeTriplet<VD,ED> threadTriplet = edgeTriplets[tid];
                    int finalTid = tid;
                    executorService.execute(
                        () -> {
                            while (true) {
                                int curBegin =
                                    Math.min(atomicInteger.getAndAdd(chunkSize), originEnd);
                                int curEnd = Math.min(curBegin + chunkSize, originEnd);
                                if (curBegin >= originEnd) {
                                    break;
                                }
                                try {
                                    for (long lid = curBegin; lid < curEnd; ++lid) {
                                        if (inComingMessageStore.messageAvailable(lid)) {
                                            threadTriplet.setSrcOid(idManager.lid2Oid(lid), vertexDataManager.getVertexData(lid));
                                            edgeManager.iterateOnEdgesParallel(finalTid, lid, threadTriplet, sendMsg, outgoingMessageStore);
                                        }
                                    }
                                } catch (Exception e) {
                                    e.printStackTrace();
                                }
                            }
                            countDownLatch.countDown();
                        });
                }
                try {
                    countDownLatch.await();
                } catch (Exception e) {
                    e.printStackTrace();
                    executorService.shutdown();
                }
            }

            //FIXME: flush message
            outgoingMessageStore.flushMessage(messageManager);
        } else {
            logger.info("Frag {} No message received", graphxFragment.fid());
        }
    }

    public void postApp() {
        logger.info("Post app");
    }

    /**
     * To receive message from grape, we need some wrappers. double -> DoubleMessage. long ->
     * LongMessage
     *
     * @param receiveVertex vertex
     * @return true if message received.
     */
    private boolean receiveMessage(Vertex<Long> receiveVertex) {
        int msgReceived = 0;
        //receive message
        if (conf.getEdataClass().equals(Double.class) || conf.getEdataClass()
            .equals(double.class)) {
            DoubleMsg msg = FFITypeFactoryhelper.newDoubleMsg();
            while (messageManager.getMessage(graphxFragment, receiveVertex, msg)) {
                //logger.info("frag {} get message: {}, {}", graphxFragment.fid(), receiveVertex.GetValue(), msg.getData());
                inComingMessageStore.addLidMessage(receiveVertex.GetValue(),
                    (MSG_T) (Double) msg.getData());
                msgReceived += 1;
            }
        } else if (conf.getEdataClass().equals(Long.class) || conf.getEdataClass()
            .equals(long.class)) {
            LongMsg msg = FFITypeFactoryhelper.newLongMsg();
            while (messageManager.getMessage(graphxFragment, receiveVertex, msg)) {
                //logger.info("frag {} get message: {}, {}", graphxFragment.fid(), receiveVertex.GetValue(), msg.getData());
                inComingMessageStore.addLidMessage(receiveVertex.GetValue(),
                    (MSG_T) (Long) msg.getData());
                msgReceived += 1;
            }
        } else {
            logger.info("Not supported msg type");
        }
        logger.info("frag {} received msg from others {}", graphxFragment.fid(), msgReceived);
        return msgReceived > 0;
    }

    public VertexDataManager<VD> getVertexDataManager(){
        return vertexDataManager;
    }

    public MessageStore<MSG_T,VD> getOutgoingMessageStore(){
        return outgoingMessageStore;
    }
}

File: analytical_engine/java/graphx-on-graphscope/src/main/scala/org/apache/spark/graphx/impl/graph/GraphXProxy.scala
package org.apache.spark.graphx.impl.graph


import com.alibaba.graphscope.ds.Vertex
import com.alibaba.graphscope.fragment.IFragment
import com.alibaba.graphscope.graphx.{GSEdgeTriplet, SharedMemoryRegistry}
import com.alibaba.graphscope.parallel.DefaultMessageManager
import com.alibaba.graphscope.utils.{FFITypeFactoryhelper, MappedBuffer, TypeUtils}
import org.apache.spark.graphx.impl.message.DefaultMessageStore
import org.apache.spark.graphx.traits.{MessageStore, VertexDataManager, GraphXVertexIdManager}
import org.apache.spark.graphx.{EdgeTriplet, GraphXConf, GraphXFactory, VertexId}
import org.apache.spark.internal.Logging

import java.util.Objects
import java.util.concurrent.{CountDownLatch, ExecutorService, Executors}
import java.util.concurrent.atomic.AtomicInteger
import scala.collection.Iterator
import scala.reflect.ClassTag

class GraphXProxy[VD : ClassTag, ED : ClassTag, MSG_T: ClassTag](val conf: GraphXConf[VD, ED],
                                                       val fragment : IFragment[Long,Long,_,_],
                                                       val messageManager: DefaultMessageManager,
                                                       val initialMessage: MSG_T,
                                                       val vprog: (VertexId, VD, MSG_T) => VD,
                                                       var sendMsg:EdgeTriplet[VD, ED] => Iterator[(VertexId, MSG_T)],
                                                       var mergeMsg: (MSG_T,MSG_T) => MSG_T,
                                                       val maxIterations: Int,
                                                       var numCores: Int, val vdataPath: String, val vdataSize: Long) extends Logging{
  private var round = 0
  private var vprogTime = 0L
  private var msgSendTime = 0L
  private var receiveTime = 0L
  private var flushTime = 0L
  private val vertexChunkSize = 4096
  private val edgeChunkSize = 1024

  val vdataBuffer: MappedBuffer = SharedMemoryRegistry.getOrCreate.mapFor(vdataPath, vdataSize)
  log.info(s"mapped vdata buffer ${vdataBuffer} of size ${vdataSize}")

  val executorService: ExecutorService = Executors.newFixedThreadPool(numCores)

  val idManager: GraphXVertexIdManager = GraphXFactory.createVertexIdManager(conf, fragment)
  val vertexDataManager: VertexDataManager[VD] = GraphXFactory.createVertexDataManager[VD,ED](conf, fragment)
  val inComingMessageStore: DefaultMessageStore[VD,MSG_T] = GraphXFactory.createDefaultMessageStore[VD,MSG_T](conf, fragment, idManager, vertexDataManager,mergeMsg)
  val outgoingMessageStore: DefaultMessageStore[VD,MSG_T] = GraphXFactory.createDefaultMessageStore[VD,MSG_T](conf,fragment, idManager, vertexDataManager,mergeMsg)

    val edgeTriplets = new Array[GSEdgeTriplet[VD, ED]](numCores)
    for (i <- 0 until numCores) {
      this.edgeTriplets(i) = GraphXFactory.createEdgeTriplet(conf)
    }
    val edgeManager: EdgeManagerImpl[VD, ED] = GraphXFactory.createEdgeManager(conf,fragment, idManager, vertexDataManager, numCores)


  def ParallelPEval(): Unit = {
    vprogTime -= System.nanoTime

    val innerVerticesNum = fragment.getInnerVerticesNum.toInt
    val vprogInteger = new AtomicInteger(0)
    val vprogLatch = new CountDownLatch(numCores)
    for (tid <- 0 until numCores) {
      executorService.execute(new PEvalVprog(tid, vprogInteger, innerVerticesNum, vprogLatch))
    }
    try{
      vprogLatch.await()
    }
    catch {
      case e: Exception =>
        e.printStackTrace()
        executorService.shutdown()
    }
    vprogTime += System.nanoTime

    /*--------------send msg ----------------*/
    msgSendTime -= System.nanoTime
    val msgSendInteger = new AtomicInteger(0)
    val msgSendLatch = new CountDownLatch(numCores)
    for (tid <- 0 until numCores) {
      val threadTriplet = edgeTriplets(tid)
      executorService.execute(new PEvalSendMsg(tid, msgSendInteger, innerVerticesNum, msgSendLatch, threadTriplet))
    }
    try{
      msgSendLatch.await()
    }
    catch {
      case e: Exception =>
        e.printStackTrace()
        executorService.shutdown()
    }
    msgSendTime += System.nanoTime

    /*--------------flush msg ----------------*/
    flushTime -= System.nanoTime
    outgoingMessageStore.flushMessage(messageManager)
    flushTime += System.nanoTime
    //messages to self are cached locally.
    round = 1
  }
  class PEvalVprog(val tid : Int, val atomicInteger: AtomicInteger, ivNum : Int, countDownLatch: CountDownLatch) extends Runnable{
    override def run(): Unit = {
      while (true){
        val curBegin = Math.min(atomicInteger.getAndAdd(vertexChunkSize), ivNum)
        val curEnd = Math.min(curBegin + vertexChunkSize, ivNum)
        if (curBegin >= curEnd){
          countDownLatch.countDown()
          return
        }
        var lid = curBegin
        while (lid < curEnd) {
          vertexDataManager.setVertexData(lid, vprog.apply(idManager.lid2Oid(lid), vertexDataManager.getVertexData(lid), initialMessage))
          lid += 1
        }
      }
    }
  }
  class PEvalSendMsg(tid: Int, atomicInteger: AtomicInteger, ivNum: Int, latch: CountDownLatch, triplet: GSEdgeTriplet[VD,ED]) extends Runnable{
    override def run(): Unit = {
      while (true) {
        val curBegin = Math.min(atomicInteger.getAndAdd(edgeChunkSize), ivNum)
        val curEnd = Math.min(curBegin + edgeChunkSize, ivNum)
        if (curBegin >= curEnd){
          latch.countDown()
          return
        }
        var lid = curBegin
        while (lid < curEnd) {
          triplet.setSrcOid(idManager.lid2Oid(lid), vertexDataManager.getVertexData(lid))
          edgeManager.iterateOnEdgesParallel(tid, lid, triplet, sendMsg, outgoingMessageStore)
          lid += 1
        }
      }
    }
  }

  def ParallelIncEval: Boolean = {
    receiveTime -= System.nanoTime
    if (round >= maxIterations) {
      return true
    }
    val receiveVertex : Vertex[Long] = FFITypeFactoryhelper.newVertexLong.asInstanceOf[Vertex[Long]]
    val innerVerticesNum = this.fragment.getInnerVerticesNum.toInt

    inComingMessageStore.clear()
    outgoingMessageStore.swap(inComingMessageStore)
    val outerMsgReceived = receiveMessage(receiveVertex.asInstanceOf[Vertex[java.lang.Long]])
    outgoingMessageStore.clear()
    receiveTime += System.nanoTime

    if (outerMsgReceived || inComingMessageStore.hasMessages) {
      vprogTime -= System.nanoTime
      val vprogInteger = new AtomicInteger(0)
      val vprogLatch = new CountDownLatch(numCores)

      for (tid <- 0 until numCores){
        executorService.execute(new IncEvalVprog(tid, vprogInteger,innerVerticesNum, vprogLatch))
      }
      try{
        vprogLatch.await()
      }
      catch {
        case e: Exception =>
          e.printStackTrace()
          executorService.shutdown()
      }
      vprogTime += System.nanoTime
      /*---------------send msg ----------------*/
      msgSendTime -= System.nanoTime
      val msgSendInteger = new AtomicInteger(0)
      val msgSendLatch = new CountDownLatch(numCores)
      for (tid <- 0 until numCores) {
        val threadTriplet = edgeTriplets(tid)
        executorService.execute(new IncEvalSendMsg(tid, msgSendInteger, innerVerticesNum, msgSendLatch, threadTriplet))
      }
      try{
        msgSendLatch.await()
      }
      catch {
        case e: Exception =>
          e.printStackTrace()
          executorService.shutdown()
      }
      msgSendTime += System.nanoTime
      flushTime -= System.nanoTime
      //FIXME: flush message
      outgoingMessageStore.flushMessage(messageManager)
      flushTime += System.nanoTime
    }
    else {
      log.info(s"Frag [${fragment.fid()}] No message received")
    }
    round += 1
    log.info(s"[frag ${fragment.fid()} Profiling]: end of round ${round}, receiveMsg cost ${receiveTime / 1000000}ms, vprog cost ${vprogTime / 1000000}ms, sendMsg cost ${ msgSendTime / 1000000} flush msg ${flushTime / 1000000}")
    false
  }

  class IncEvalVprog(val tid : Int, val atomicInteger: AtomicInteger, ivNum : Int, countDownLatch: CountDownLatch) extends Runnable{
    override def run(): Unit = {
      while (true){
        val curBegin = Math.min(atomicInteger.getAndAdd(vertexChunkSize), ivNum)
        val curEnd = Math.min(curBegin + vertexChunkSize, ivNum)
        if (curBegin >= curEnd){
          countDownLatch.countDown()
          return
        }
        var lid = curBegin
        while (lid < curEnd) {
          if (inComingMessageStore.messageAvailable(lid)) {
            val oid = idManager.lid2Oid(lid)
            val vdata = vertexDataManager.getVertexData(lid)
            val msg = inComingMessageStore.getMessage(lid)
            vertexDataManager.setVertexData(lid, vprog.apply(oid, vdata, msg))
          }
          lid += 1
        }
      }
    }
  }
  class IncEvalSendMsg(tid: Int, atomicInteger: AtomicInteger, ivNum: Int, latch: CountDownLatch, triplet: GSEdgeTriplet[VD,ED]) extends Runnable{
    override def run(): Unit = {
      while (true) {
        val curBegin = Math.min(atomicInteger.getAndAdd(edgeChunkSize), ivNum)
        val curEnd = Math.min(curBegin + edgeChunkSize, ivNum)
        if (curBegin >= curEnd){
          latch.countDown()
          return
        }
        var lid = curBegin
        while (lid < curEnd) {
          if (inComingMessageStore.messageAvailable(lid)) {
            triplet.setSrcOid(idManager.lid2Oid(lid), vertexDataManager.getVertexData(lid))
            edgeManager.iterateOnEdgesParallel(tid, lid, triplet, sendMsg, outgoingMessageStore)
          }
          lid += 1
        }
      }
    }
  }

  def persistVdata(): Unit = {
    val innerVertexNum = fragment.getInnerVerticesNum.toInt
    val bytesToWrite = innerVertexNum * (8 + TypeUtils.classToBytes(conf.getVdClass))
    if (Objects.nonNull(vdataBuffer) && vdataBuffer.remaining >= bytesToWrite) {
      log.info("Writing vertex data to graphx shared memory")
      vdataBuffer.writeLong(bytesToWrite)
      if (conf.getVdClass.equals(classOf[Long]) || conf.getVdClass.equals(classOf[java.lang.Long])) {
        for (lid <- 0 until innerVertexNum) {
          vdataBuffer.writeLong(idManager.lid2Oid(lid))
          vdataBuffer.writeLong(vertexDataManager.getVertexData(lid).asInstanceOf[VD].asInstanceOf[Long])
          log.info(s"Writing ${idManager.lid2Oid(lid)}, ${vertexDataManager.getVertexData(lid)}")
        }
        log.info("Finish writing long vdata to shared memory")
      }
      else if (conf.getVdClass.equals(classOf[Double]) || conf.getVdClass.equals(classOf[java.lang.Double])) {
        for (lid <- 0 until innerVertexNum) {
          vdataBuffer.writeLong(idManager.lid2Oid(lid))
          vdataBuffer.writeDouble(vertexDataManager.getVertexData(lid).asInstanceOf[Double])
        }
        log.info("Finish write double vdata to shared memory")
      }
      else if (conf.getVdClass.equals(classOf[Int]) || conf.getVdClass.equals(classOf[java.lang.Integer])) {
        for (lid <- 0 until innerVertexNum) {
          vdataBuffer.writeLong(idManager.lid2Oid(lid))
          vdataBuffer.writeInt(vertexDataManager.getVertexData(lid).asInstanceOf[VD].asInstanceOf[Integer])
        }
        log.info("Finish write int vdata to shared memory")
      }
      else log.error("Not recognized vdata class: " + conf.getVdClass)
    }
    else log.error(s"vdata null or space not enough ${vdataBuffer}, remaining ${vdataBuffer.remaining}, to write {bytesToWrite}")
  }

  def postApp(): Unit = {
    log.info("Post app")
  }

  /**
   * To receive message from grape, we need some wrappers. double -> DoubleMessage. long ->
   * LongMessage
   *
   * @param receiveVertex vertex
   * @return true if message received.
   */
  private def receiveMessage(receiveVertex: Vertex[java.lang.Long]) : Boolean = {
    var msgReceived = 0
    //receive message
    if (conf.getEdClass.equals(classOf[Double]) || conf.getEdClass.equals(classOf[java.lang.Double])) {
      val msg = FFITypeFactoryhelper.newDoubleMsg
      while ( {
        messageManager.getMessage(fragment, receiveVertex, msg)
      }) { //logger.info("frag {} get message: {}, {}", graphxFragment.fid(), receiveVertex.GetValue(), msg.getData());
        inComingMessageStore.addLidMessage(receiveVertex.GetValue, msg.getData.asInstanceOf[Double].asInstanceOf[MSG_T])
        msgReceived += 1
      }
    }
    else if (conf.getEdClass.equals(classOf[Long]) || conf.getEdClass.equals(classOf[java.lang.Long])) {
      val msg = FFITypeFactoryhelper.newLongMsg
      while ( {
        messageManager.getMessage(fragment, receiveVertex, msg)
      }) {
        inComingMessageStore.addLidMessage(receiveVertex.GetValue, msg.getData.asInstanceOf[MSG_T])
        msgReceived += 1
      }
    }
    else log.info("Not supported msg type ${conf.getEdClass.getName}")
    log.info(s"Frag [${fragment.fid}] received msg from others ${msgReceived}")
    msgReceived > 0
  }

  def getVertexDataManager: VertexDataManager[VD] = vertexDataManager

  def getOutgoingMessageStore: MessageStore[MSG_T] = outgoingMessageStore
}

